pkg/template/template.go

```diff
@@ -60,24 +57,59 @@ func (p *Processor) Process(template *api.Template) field.ErrorList {
 			}
 			item = decodedObj
 		}
-
-		newItem, err := p.SubstituteParameters(paramMap, item)
-		if err != nil {
-			templateErrors = append(templateErrors, field.Invalid(idxPath.Child("parameters"), template.Parameters, err.Error()))
-		}
 		// If an object definition's metadata includes a namespace field, the field will be stripped out of
 		// the definition during template instantiation.  This is necessary because all objects created during
 		// instantiation are placed into the target namespace, so it would be invalid for the object to declare
-		//a different namespace.
-		stripNamespace(newItem)
-		if err := util.AddObjectLabels(newItem, template.ObjectLabels); err != nil {
+		// a different namespace.
+		stripNamespace(item)
+		if err := util.AddObjectLabels(item, template.ObjectLabels); err != nil {
 			templateErrors = append(templateErrors, field.Invalid(idxPath.Child("labels"),
 				template.ObjectLabels, fmt.Sprintf("label could not be applied: %v", err)))
 		}
-		template.Objects[i] = newItem
+		template.Objects[i] = item
+	}
+	if fieldError := p.GenerateParameterValues(template); fieldError != nil {
+		templateErrors = append(templateErrors, fieldError)
+	}
+
+	// accrued errors from processing the template objects and parameters
+	if len(templateErrors) != 0 {
+		return templateErrors, nil
 	}
 
-	return templateErrors
+	// Place parameters into a map for efficient lookup
+	paramMap := make(map[string]api.Parameter)
+	for _, param := range template.Parameters {
+		paramMap[param.Name] = param
+	}
+
+	// Turn the template object into json so we can do search/replace on it
+	// to substitute parameter values.
+	serializer, found := kapi.Codecs.SerializerForMediaType("application/json", nil)
+	if !found {
+		return templateErrors, errors.New("Could not load json serializer")
+	}
+
+	templateBytes, err := runtime.Encode(serializer, template)
```

Comment:
We can't use regex over json unfortunately, there are to many ways it can fail.  Going to have to transform in place over the unstructured scheme.